### PR TITLE
feat: Add derivation for `risc0-groth16-prover` binaries

### LINKED ISSUE #28: `patchelf` dynamically linked prover binary
NixOS is unable to run dynamically linked executables by default. The docker container that is run in `setup.sh` produces some binary that is placed in the `relay` directory and is called via path by bonsol. The best way for this to be made available to nix is to have nix build the image and copy the paths into the nix store, and make them available to bonsol. This seems to be the only blocker for running a complete NixOS build with bonsol, though everything is working on non-NixOS linux when using Nix since it is able to run dynamically linked executables. Examples of how to use the `autoPetchelfHook` are already abundant in the bonsol codebase, and the official [documentation on using docker](https://nixos.wiki/wiki/Docker) with nix is great.

### LINKED ISSUE #40: Use cachix binary cache in CI
Currently we use the magic nix cache which is great, however it can sometimes take a while for CI to complete the first time. We can potentially speed this up by using the cachix binary cache. We could also use CI to update the binary cache automatically, after a pull request is merged into main.

### PR DESCRIPTION
**Pull request by eureka-cpu**
Closes #28 
Closes #40 

Adds the groth16 prover binaries as statically linked binaries in the nix store, allowing them to be used in a sandbox environment. This partially unblocks the use case for NixOS systems, as it removes the need for the prover binaries to be dynamically linked.

This also adds the bonsol cache to CI for the nix checks, which speeds up builds from taking 40m to under 2m in most cases.

### CODE REVIEW

**Comment by eureka-cpu** [comment]
There were also some path resolution things that I missed from #39

**Comment by eureka-cpu** [comment]
> There were also some path resolution things that I missed from #39 which may be why the vercel deployment failed?

The cause was that the `pnpm-workspace.yaml` still uses `channel` instead of the new `bonsol` naming convention (not that it's super important) but it causes the vercel deployment to fail.

File: docs/docs/contributing/contributing.md [hunk 1]
```diff
@@ -20,7 +20,7 @@ that avoids unecessary downloading, rebuilding and patching. With one command yo
 
 The following link will install nix with the above features and include the bonsol binary cache as a trusted substitutor. Without the substitutor many dependencies will build from source, which could take a the first time they are built!
 ```bash
-curl --proto '=https' --tlsv1.2 -sSf -L https://install.determinate.systems/nix | sh -s -- install --extra-conf "extra-substitutors = https://bonsol.cachix.org" --extra-conf "extra-trusted-public-keys = bonsol.cachix.org-1:yz7vi1rCPW1BpqoszdJvf08HZxQ/5gPTPxft4NnT74A="
+curl --proto '=https' --tlsv1.2 -sSf -L https://install.determinate.systems/nix | sh -s -- install --extra-conf "extra-substituters = https://bonsol.cachix.org" --extra-conf "extra-trusted-public-keys = bonsol.cachix.org-1:yz7vi1rCPW1BpqoszdJvf08HZxQ/5gPTPxft4NnT74A="
```
Comment (jac18281828) [inline docs/docs/contributing/contributing.md:23]: is substituters a nix thing?

**Review by jac18281828 — APPROVED** [review]
Looks good, checks are passing.   🚀

**Pull request merged**